Make bootstrap.py compatible with Puppet 4

h4xr opened the pull request:
Fixes #229 

Signed-off-by: Saurabh Badhwar <sbadhwar@redhat.com>

Commit:
Make bootstrap.py compatible with Puppet 4

Signed-off-by: Saurabh Badhwar <sbadhwar@redhat.com>

evgeni commented on bootstrap.py:
```diff
@@ -383,7 +383,10 @@ def install_puppet_agent():
     enable_service("puppet")
 
-    puppet_conf = open('/etc/puppet/puppet.conf', 'wb')
+    if os.path.isfile('/etc/puppet/puppet.conf'):
```
can we turn that around and handle `/etc/puppetlabs` as the special case? feels slightly "cleaner" to me

evgeni commented on bootstrap.py:
```diff
@@ -407,7 +410,10 @@ def install_puppet_agent():
     print_generic("Visit the UI and approve this certificate via Infrastructure->Capsules")
     print_generic("if auto-signing is disabled")
-    exec_failexit("/usr/bin/puppet agent --test --noop --tags no_such_tag --waitforcert 10")
+    if os.path.isfile('/usr/bin/puppet'):
```
can we turn that around and test for `/opt/puppetlabs/puppet/bin/puppet`? and in the else case, run puppet without a path at all, so the shell can find it even in another folder?

Commit:
Allow puppet to be configured according to the detected version

The changes add functionality for detecting the installed version
of puppet and configures the puppet accordingly.

Signed-off-by: Saurabh Badhwar <sbadhwar@redhat.com>

evgeni commented on bootstrap.py:
```diff
@@ -218,6 +218,45 @@ def install_prereqs():
+    return executable_path
+
+
+def get_puppet_version():
```
I think I'd still prefer using the RPMdb like this:

```python
def get_puppet_version():
    ts = rpm.TransactionSet()
    for name in ['puppet', 'puppet-agent']:
        mi = ts.dbMatch('name', name)
        for h in mi:
            puppet_version = int(h['version'].split('.')[0])
            if h['name'] == 'puppet-agent' and puppet_version == 1:
                puppet_version = 4
            return puppet_version
```

This works fine for Puppet 3, 4 and 5

@sideangleside whatcha think?

h4xr replied:
@evgeni If I am not getting wrong, we are comparing the puppet-agent packaging version here. I doubt if this will be future proof or not, in case packaging convention changes in future for the newer releases of Puppet.

evgeni replied:
correct, it's using the packaging version. I would not expect the major version of the packaging to differ from the major version of the agent in the future (well, puppet4 agent is an exemption here :()

sideangleside replied:
I am strongly supportive of checking the RPMs version number as it is consistent with other checks for packages that we do. Additionally, checking via an executable's location is just as prone to packaging changes as anything else, so i'd want to be consistent.

evgeni commented on bootstrap.py:
```diff
@@ -383,7 +422,15 @@ def install_puppet_agent():
+    puppet_major_version = get_puppet_version()
+    if puppet_major_version == 3:
+        puppet_conf_file = '/etc/puppet/puppet.conf'
+    elif puppet_major_version == 4:
```
can we make this `puppet_major_version in [4,5]`? as Puppet5 uses the same path and that would make it more future proof

sideangleside replied:
👍

evgeni commented on bootstrap.py:
```diff
@@ -218,6 +218,45 @@ def install_prereqs():
         delete_file('/etc/yum.repos.d/katello-client-bootstrap-deps.repo')
 
 
+def get_puppet_path():
```
this would be then (if we use the RPMdb)

```python
puppet_major_version = get_puppet_version()
if puppet_major_version == 3:
    return '/usr/bin/puppet'
elif puppet_major_version in [4,5]:
    return '/opt/puppetlabs/puppet/bin/puppet'
else:
    error

Commit:
Use RPMdb for determining Puppet version

Signed-off-by: Saurabh Badhwar <sbadhwar@redhat.com>

Commit:
Adapt puppet configuration for different versions of puppet

Signed-off-by: Saurabh Badhwar <sbadhwar@redhat.com>

evgeni commented on bootstrap.py:
```diff
@@ -383,13 +414,35 @@ def install_puppet_agent():
+rundir = /var/run/puppet
+ssldir = /etc/puppetlabs/puppet/ssl
+""")
+    else:
```
that code is unreached, we already verified that it's version 3, 4 or 5 when looking for the conf path

evgeni commented on bootstrap.py:
```diff
@@ -383,13 +414,35 @@ def install_puppet_agent():
+        print_error("Unsupported puppet version")
+        sys.exit(1)
+    puppet_conf = open(puppet_conf_file, 'wb')
+    if puppet_major_version == 3:
```
I think instead of having another set of `if/elif` here, I'd put the dirs (or the whole `[main]` section) into a variable during the conf path detection, and just use the variable in one `write()`.

evgeni reviewed (approved):
two comments, which would make the code a bit cleaner (IMHO), but neither is really blocking and the code works just fine as it is. ACK from me.

would love to hear a second ACK from @sideangleside before merging, though

evgeni commented on bootstrap.py:
```diff
@@ -383,13 +414,31 @@ def install_puppet_agent():
+[main]
+vardir = /opt/puppetlabs/puppet/cache
+logdir = /var/log/puppetlabs/puppet
+rundir = /var/run/puppet
```
this should be `/var/run/puppetlabs`

evgeni commented on bootstrap.py:
```diff
@@ -383,13 +414,31 @@ def install_puppet_agent():
+    puppet_major_version = get_puppet_version()
+    if puppet_major_version == 3:
+        puppet_conf_file = '/etc/puppet/puppet.conf'
+        main_section = """
```
cosmetic: this makes `main_section` start with a newline, and thus the generated `puppet.conf` starts with two newlines. can we change this to `main_section = """[main]…`?

evgeni commented on bootstrap.py:
```diff
@@ -383,13 +414,31 @@ def install_puppet_agent():
+"""
+    elif puppet_major_version in [4, 5]:
+        puppet_conf_file = '/etc/puppetlabs/puppet/puppet.conf'
+        main_section = """
```
cosmetic: this makes `main_section` start with a newline, and thus the generated `puppet.conf` starts with two newlines. can we change this to `main_section = """[main]…`?

evgeni reviewed (changes requested):
one functional, and two cosmetic comments, good to go when at least the functional is addressed.

evgeni reviewed (approved):
ACK

waiting on @sideangleside
